### PR TITLE
mgr/cephadm: Drop a bunch of spec properties that are just used for jinja2 templates

### PR DESCRIPTION
**Pull request by sebastian-philipp**
* Drop GrafanaSpec
* Drop AlertmanagerSpec
* Drop a bunch of extra properties of IngressSpec

I see three benefits of this PR:

* It removes a bunch of service specs that are only needed to pass-though props to jinja2. I.e. avoid polluting the specs
* it enables us to more easily add new features for those use cases
* It might be a way to help avoiding the helm chart fate (ballooning yaml configs)
* removes user_data from alertmanger specs.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
changlog: deleted AlertmanagerSpec

**Comment by sebastian-philipp** [comment]
changelog: Drop user_data from alertmanager

Wow, this simple change was way too complicated to build. we have to get better at this.

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 1]
```diff
@@ -152,11 +175,11 @@ def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[st
             addr = self.mgr.inventory.get_addr(dd.hostname)
             dashboard_urls.append(build_url(scheme=proto, host=addr, port=port))
 
-        context = {
-            'dashboard_urls': dashboard_urls,
-            'default_webhook_urls': default_webhook_urls
-        }
-        yml = self.mgr.template.render('services/alertmanager/alertmanager.yml.j2', context)
+        context = AlertmanagerYamlContext(
+            dashboard_urls=dashboard_urls,
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/monitoring.py:179]: What's the reason for not keeping default_webhook_urls here? I know it got moved from part of user_data to it's own thing but why not still include it in the context?

File: src/pybind/mgr/cephadm/tests/test_migration.py [hunk 2]
```diff
@@ -228,3 +228,41 @@ def test_migrate_admin_client_keyring(cephadm_module: CephadmOrchestrator):
...
+            'default_webhook_urls': [
+                'http://url1'
+            ]
+        }
+
+        cephadm_module.migration.migrate(True)
+        assert cephadm_module.migration_current == 4
+
+        cephadm_module.migration.migrate()
+        assert cephadm_module.migration_current == LAST_MIGRATION
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/tests/test_migration.py:268]: newline after this for flake8

**Review by adk3798 — COMMENTED** [review]
I think this is a good idea for allowing us to essentially allow add properties to a spec without having to explicitly have them as parameters. There are just a few things I'm confused about on a technical level, notably how things from other_properties are propagated into a template and how webhook urls are getting set for alertmanager (although maybe the answer to the first question also answers the second)

File: src/pybind/mgr/cephadm/templates/services/ingress/haproxy.cfg.j2 [hunk 3]
```diff
@@ -6,15 +6,15 @@ global
...
+{% if ssl_cert %}
+  {% if ssl_dh_param %}
+    tune.ssl.default-dh-param {{ ssl_dh_param }}
   {% endif %}
-  {% if spec.ssl_ciphers %}
-    ssl-default-bind-ciphers {{ spec.ssl_ciphers | join(':') }}
+  {% if ssl_ciphers %}
+    ssl-default-bind-ciphers {{ ssl_ciphers | join(':') }}
   {% endif %}
-  {% if spec.ssl_options %}
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/templates/services/ingress/haproxy.cfg.j2:16]: exactly. I gather all the fields that are used and subtract the known fields. The result is a list of fields that are used without being defined.  That's why I replaced `{% if spec.ssl_options %}` with `{% if ssl_options %}`. 

WDYT? Is this a sane thing to do?

File: src/python-common/ceph/deployment/service_spec.py [hunk 4]
```diff
@@ -429,10 +430,9 @@ def _cls(cls: Type[ServiceSpecT], service_type: str) -> Type[ServiceSpecT]:
             'nfs': NFSServiceSpec,
             'osd': DriveGroupSpec,
             'iscsi': IscsiServiceSpec,
-            'alertmanager': AlertManagerSpec,
+            'alertmanager': MonitoringSpec,
             'ingress': IngressSpec,
             'container': CustomContainerSpec,
-            'grafana': GrafanaSpec,
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/service_spec.py:435]: whoops!

File: src/python-common/ceph/deployment/service_spec.py [hunk 5]
```diff
@@ -501,6 +503,10 @@ def __init__(self,
         if config:
             self.config = {k.replace(' ', '_'): v for k, v in config.items()}
 
+        #: Dict of properties that can be used in Jinja2 templates without bein
+        #: explicitly tracked in the spec files.
+        self.other_properties = other_properties or {}
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/service_spec.py:507]: haproxy, alertmanager, grafana!

File: src/pybind/mgr/cephadm/services/monitoring.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/services/monitoring.py:179]: my resoning was: we're not doing anything with it. we're just passing it to jinja2. That's why I removed it completely from the python code base

File: src/pybind/mgr/cephadm/templates/services/ingress/haproxy.cfg.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/templates/services/ingress/haproxy.cfg.j2:16]: So, could you set one of these fields by passing it as one of the "other properties". If so, I get the idea. Explicitly support a handful of necessary fields in the spec then cover everything else with the other properties section. Then when we want to "support" something else new we can just add a new piece to the template only so it's still doable without a custom template but our specs don't grow forever. If that's how it works I like it.

[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/templates/services/ingress/haproxy.cfg.j2:16]: exactly!

**Pull request closed without merging**